### PR TITLE
Add MAX_STATEMENT_TIME resource limit

### PR DESCRIPTION
**Pull request by mstinsky**
This MR adds MAX_STATEMENT_TIME to resource_limits as already requested in #211.

MAX_STATEMENT_TIME is supported since mariadb 10.1.1, as resource_limits are only supported by this module for maiadb >= 10.2 there should not be a requirement to add a new version check.

I also refactored the `server_supports_alter_user` function as it was not using the same implementations/impl paradigma as other versions checks, which confused me quite a bit while searching for the version check.

### CODE REVIEW

**Comment by mstinsky** [comment]
Overlooked that the parameter ist called different in mysql and mariadb, will update the PR soon.

**Comment by Andersson007** [comment]
@mstinsky hello, thanks for the PR!
Could you please:
- Add a changelog [fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) of `minor_changes` type
- Can we add integration tests to somewhere in `tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml` ?

thank you

**Comment by mstinsky** [comment]
hey @Andersson007,

sure I wanted to add some integration and unit tests. The problem at the moment is that I am not able to get the integration tests working locally.
I followed the guide from https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html on a fresh ubuntu 20 test vm, but when I execute `ansible-test integration test_mysql_user --docker -v` the test fails with:
```
TASK [setup_controller : setup_controller | Setvars | Set Fact] ****************
fatal: [testhost]: FAILED! => {"msg": "The 'file' lookup had an issue accessing the file '/root/ansible_collections/community/mysql/tests/integration/connector_name'. file not found, use -vvvvv to see paths searched"}
```

At the moment I am not really sure how to fix that.

**Comment by Andersson007** [comment]
@mstinsky 
>At the moment I am not really sure how to fix that.

There's CI overhaul in progress, we'll switch to another scheme.
For now, you can just push your commits with tests here.

If you're curious, there's https://github.com/ansible-collections/community.mysql/blob/main/TESTING.md with info how to run them locally but don't focus on this at the moment as it needs polishing, you can figure out after this PR is merged.
We'd be happy for any feedback on testing here (in issues/discussions that are not created yet) or in the [mysql](https://matrix.to/#/#mysql:ansible.com) room on [Matrix](https://docs.ansible.com/ansible/devel/community/communication.html#ansible-community-on-matrix)

**Comment by Andersson007** [comment]
and please ignore "codecov" errors

File: changelogs/fragments/523-add-max_statement_time_resource-limit.yml
Comment (Andersson007) [inline changelogs/fragments/523-add-max_statement_time_resource-limit.yml]:
```suggestion
  - mysql_user - add ``MAX_STATEMENT_TIME`` support to the ``resource_limits`` argument (https://github.com/ansible-collections/community.mysql/issues/211).
```

Comment (mstinsky) [reply changelogs/fragments/523-add-max_statement_time_resource-limit.yml]: Changed.

**Comment by mstinsky** [comment]
After fiddling around a bit it seems I got the mariadb part working with tests.
As I am not really familiar with mysql it took me a while to figure out that it seems that mysql does not support setting execution time for a user. Therefore my approach so far to set `max_statement_time` for mariadb and `max_execution_time` for mysql will not work at all.

I'll need to rework the PR in the next days so that the resource_limit for max_statement_time is only available for mariadb, after that the PR should be good I hope :)

**Comment by Andersson007** [comment]
> After fiddling around a bit it seems I got the mariadb part working with tests. As I am not really familiar with mysql it took me a while to figure out that it seems that mysql does not support setting execution time for a user. Therefore my approach so far to set `max_statement_time` for mariadb and `max_execution_time` for mysql will not work at all.
> 
> I'll need to rework the PR in the next days so that the resource_limit for max_statement_time is only available for mariadb, after that the PR should be good I hope :)

@mstinsky the plan SGTM. Please reflect the fast that it works only for mariadb in docs. Also you can not to fail in case of mysql but ignore (in this case this should be also reflected in docs). Thanks

There's one sanity error:
```
ERROR: Found 1 pylint issue(s) which need to be resolved:
ERROR: tests/unit/plugins/module_utils/test_mysql_user_implementation.py:8:0: unused-import: Unused use_max_statement_time imported from ansible_collections.community.mysql.plugins.module_utils.implementations.mysql.user (50%)
```
You can still run sanity/unit tests locally as described in the Quick start guide

**Comment by mstinsky** [comment]
@Andersson007 I think the PR is now in a working state, at least from my point of view.

The last thing that I dont understand right now is where the sanity test error for validating the module is coming from (https://github.com/ansible-collections/community.mysql/actions/runs/4607850899/jobs/8142888629?pr=523#step:3:1305)
When I execute `ansible-test sanity --test validate-modules` locally I get `exit 0`.
Maybe you can point me in the right direction to fix this? :)

All other tests + the new ones should work now.

**Comment by Andersson007** [comment]
@mstinsky sure:
- as we can see the the sanity checks fail in `devel`, so probably you use another branch (or you need to pull the latest changes to `devel` from upstream) when running the tests.
  1. clone https://github.com/ansible/ansible
  2. `cd` to it
  3. run `source hacking/env-setup`
  4. go to `~/ansible_collections/community/mysql` and run 
 
Now i see that the error is unrelated to your PR and it probably relates to https://github.com/ansible-collections/news-for-maintainers/issues/41. I.e. with the new version of Ansible, new checks were introduced.
Your PR is OK now. We can fix the sanity check later in a separate PR.

File: plugins/modules/mysql_user.py
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
        C(MAX_CONNECTIONS_PER_HOUR: num), C(MAX_USER_CONNECTIONS: num), C(MAX_STATEMENT_TIME: num) (supported only for MariaDB since collection version 3.7.0)."
```

File: plugins/module_utils/user.py
Comment (Andersson007) [inline plugins/module_utils/user.py]:
```suggestion
            module.fail_json(msg="MAX_STATEMENT_TIME resource limit is only supported by MariaDB.")
```

**Review by Andersson007 — COMMENTED** [review]
Generally LGTM (i like you use `impl`)
Only a couple of minor formatting things.
Thanks!

File: plugins/modules/mysql_user.py
Comment (Andersson007) [reply plugins/modules/mysql_user.py]: I'm also OK to have it in a separate line as in your version (line 149). In this case, i would
- Remove the choice from line 148 not to repeat it
- Add that it's supported since collection version `3.7.0` to line 149

Up to you

File: plugins/module_utils/user.py
Comment (mstinsky) [reply plugins/module_utils/user.py]: Changed.

File: plugins/modules/mysql_user.py
Comment (mstinsky) [reply plugins/modules/mysql_user.py]: I like yours better. Changed it to your version.

Comment (Andersson007) [reply plugins/modules/mysql_user.py]: thanks!

**Review by Andersson007 — APPROVED** [review]
LGTM, thanks!

@laurent-indermuehle @betanummeric would you like to take a look? If you're OK, feel free to merge of course (please ignore the sanity errors, they are unrelated)

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      check_mode: true
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
    - name: Resource limits | Assert that create user with max_statement_time is changed for MariaDB
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
    - name: Resource limits | Assert that create user with max_statement_time is failed for MySQL
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_query:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_resource_limits.yml]:
```suggestion
      community.mysql.mysql_query:
```

**Review by laurent-indermuehle — APPROVED** [review]
Great work! And great addition to the collection. I'm really impressed by your contribution.
I know I'm annoying with my suggestions about yes vs true and FQCN, but I hope to enable ansible-lint on this repo someday.

**Comment by mstinsky** [comment]
Thanks for the review @laurent-indermuehle I added your suggested changes.

**Review by Andersson007 — APPROVED** [review]

**Comment by betanummeric** [comment]
LGTM.
The module is case-sensitive, while MariaDB is not. Maybe it's convenient to take keys in `resource_limits` case-insensitive by adding `.upper()` somewhere in the module.

**Comment by mstinsky** [comment]
I force pushed a last change from me as I saw that I copy pasted the wrong resource limit when testing to change the new max_statement_time resource limit.

@betanummeric regarding your comment. If I understand you correctly you would like to have case insensitivity for all resource limits when the module adds them to mariadb. From my point of view that would be better placed in a new PR.

**Comment by betanummeric** [comment]
@mstinsky Yes, case-sensitivity should be equal for all resource limits. It's just an idea, we don't need to implement it (now).

**Comment by Andersson007** [comment]
@joekeilty-oub thanks much for reporting the issue!
@mstinsky thanks for the implementation, great job!
@laurent-indermuehle @betanummeric thanks for reviewing!

**Comment by Andersson007** [comment]
Can we close the [issue](https://github.com/ansible-collections/community.mysql/issues/211) or there's anything else left to implement?

**Comment by mstinsky** [comment]
@Andersson007 from my point of view the feature requested from https://github.com/ansible-collections/community.mysql/issues/211 is implemented so the issue can be closed.

I also wanted to thank you guys for the rapid replies on this PR, very much appriciated.

**Comment by Andersson007** [comment]
> @Andersson007 from my point of view the feature requested from #211 is implemented so the issue can be closed.
> 
> I also wanted to thank you guys for the rapid replies on this PR, very much appriciated.

@mstinsky sure, fast feedback is important we believe:) You're welcome and we'd be happy to see more contributions from you here or everywhere, e.g. in [community.postgresql](https://github.com/ansible-collections/community.postgresql):)

**Pull request merged**
